components/base/sensorcontrolled/sensorcontrolled.go

```diff
@@ -209,12 +271,130 @@ func (sb *sensorBase) isPolling() bool {
 func (sb *sensorBase) MoveStraight(
 	ctx context.Context, distanceMm int, mmPerSec float64, extra map[string]interface{},
 ) error {
-	sb.stopLoop()
 	ctx, done := sb.opMgr.New(ctx)
 	defer done()
 	sb.setPolling(false)
+	straightTimeEst := time.Duration(int(time.Second) * int(math.Abs(float64(distanceMm)/mmPerSec)))
+	startTime := time.Now()
+	timeOut := 5 * straightTimeEst
+	if timeOut < 10*time.Second {
+		timeOut = 10 * time.Second
+	}
+
+	if sb.position == nil || len(sb.conf.ControlParameters) == 0 {
+		sb.logger.CWarnf(ctx, "Position reporting sensor not available, and no control loop is configured, using base %s MoveStraight", sb.controlledBase.Name().ShortName())
+		sb.stopLoop()
+		return sb.controlledBase.MoveStraight(ctx, distanceMm, mmPerSec, extra)
+	}
+
+	initialHeading, err := sb.headingFunc(ctx)
+	if err != nil {
+		return err
+	}
+	// make sure the control loop is enabled
+	if sb.loop == nil {
+		if err := sb.startControlLoop(); err != nil {
+			return err
+		}
+	}
+
+	// initialize relevant parameters for moving straight
+	slowDownDist := calcSlowDownDist(distanceMm)
+	var initPos *geo.Point
+
+	if sb.position != nil {
+		initPos, _, err = sb.position.Position(ctx, nil)
+		if err != nil {
+			return err
+		}
+	}
+
+	ticker := time.NewTicker(time.Duration(1000./sb.controlLoopConfig.Frequency) * time.Millisecond)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ticker.C:
+			var errDist float64
+
+			angVelDes, err := sb.calcHeadingControl(ctx, initialHeading)
+			if err != nil {
+				return err
+			}
+			sb.logger.Info(angVelDes)
+			if sb.position != nil {
+				errDist, err = calcPositionError(ctx, distanceMm, initPos, sb.position)
+				if err != nil {
+					return err
+				}
+			}
+
+			if errDist < moveStraightErrTarget {
+				return sb.Stop(ctx, nil)
+			}
+
+			linVelDes := calcLinVel(errDist, mmPerSec, slowDownDist)
+			if err != nil {
+				return err
+			}
+
+			// update velocity controller
+			if err := sb.updateControlConfig(ctx, linVelDes/1000.0, angVelDes); err != nil {
```

Comment:
What's the unit conversion from `calcPosError` -> control config's SetVel ask?
We're ultimately converting to something SetPower can use right?

Reply:
the control loop handles converting things into `SetPower`. I am only calculating velocities at this layer, where `linVelDes` is `mmPerSec` from the request up until I approach the goal, where I slow down the rover. the `1000` is just converting mm to m since the control loop is configured to use the raw velocity messages from a movement sensor